### PR TITLE
Only apply CIDR policy to special identities (Makes ingress/egress policy consistent)

### PR DESCRIPTION
**Pull request by joestringer**
### Problem description

Previously, CIDR policy applied inconsistently depending on direction:
* On ingress, CIDR policies applies to all traffic
* On egress CIDR policies only applied to traffic outside the Cilium cluster prefix.

Inconsistent behaviour like this is confusing for users, so for 1.0 it would be nice to clarify the intended behaviour and make it more consistent. This PR proposes to do this by only applying them to communications with network endpoints that are not managed by Cilium.

There are multiple potential approaches to consider when making this more consistent:
1) Make CIDR apply to all traffic   
2) Make CIDR only apply to traffic between an endpoint managed by Cilium and network endpoints not managed by Cilium
2.1) Apply to all traffic outside the cluster prefix                    
2.2) Apply whenever policy cannot be determined by label-based policy.

### Apply CIDR rules everywhere

This option makes the most sense from a baseline consistency argument. Cilium policy in general is described in terms of whitelists. Either the policy specifies that something is allowed, or otherwise it is denied. At a glance, someone with a networking background that considers whether traffic matches a rule that specifies the IPv4 CIDR `0.0.0.0/0` would likely expect all traffic to match this.

The main concern with this approach is that users could apply a `0.0.0.0/0` policy and unintentionally open up all communication between nodes within their cluster. Specifically, if a user adds several policy rules that match on specific endpoints, then adds a `0.0.0.0/0` rule, then their label-based rules will now be completely ignored. This is particularly insidious if the endpoint-based rule also describes what may be allowed at L7, because the CIDR rule would override this.

### Apply CIDR rules to traffic which communicates with an external peer

This approach would instead divide possible network entities into different categories of L3:
* Those managed by a cilium with labels from container runtimes (Label-based policy)
* Those managed by something other than Cilium
  * Entity-based policy
  * CIDR-based policy

In terms of endpoints managed outside of Cilium, there has been an idea floating around the Cilium community for a while (#3553 ): Allow custom entities to be defined upon such traffic, then allow policy to be defined upon these custom entities. This idea has been documented as a roadmap feature for at least the last several months, and it would make the allow the kind of separation of concerns between policy application and traffic node IP/label mapping that we already see with label-based policy.

The policy spec v3 (#3005) was also pushing in this direction of requiring that policy is defined upon specific sets of L3 matches.

There are two variations on this approach: Using the cluster prefix to describe what is "inside" of the label-dependent policy domain (approach 2.1 above), or using the specific set of endpoints that Cilium is managing to determine what's "inside" (approach 2.2 above). All other traffic would be considered "outside", and therefore subject to CIDR policy.

Common configurations of Cilium often involve something like:
* 10.0.0.0/8 for the cluster prefix
* Container orchestrator hands out 10.x.0.0/16 subnets to individual nodes for endpoints

When using such a broad cluster prefix, particularly in the commonly-used RFC1918 address space, makes it quite possible for existing VM and/or baremetal setups to be sharing the same cluster prefix as Cilium. It may be useful to apply policy upon such subnets via CIDR, while still avoiding the potential policy pithole of adding policies that unintentionally allow communication with Cilium-managed endpoints.

### Implementation implications

If CIDR policy must be applied to all traffic, then in worst case the CIDR lookup must occur on all traffic before a drop decision is made. This applies in all cases in the BPF datapath, and only for L3-dependent L7 rules for the userspace proxies. Alternatively, with the approach proposed in this PR, the CIDR lookup can be avoided for traffic between Cilium endpoints.

#### BPF datapath

This PR implements the baseline requirements for only applying CIDR to traffic with peers not managed by Cilium. The logic is very simple - we look to see whether we have the context about the remote endpoint, and if there is a Cilium identity for it, we use that; otherwise, it is part of the "outside" and CIDR may apply. In practice this is boiled down to a check for whether the identity is a "reserved identity" or not. Reserved identities are never assigned to containers, so they represent the opposite of the "in-cluster" set of endpoints. Longer term, we could look into additional improvements on ingress by, for example, always mapping the CIDR of the packet if the identity is not known before the policy lookup and performing a single identity-based policy map lookup. This is not yet the case today; CIDR policy is implemented in a separate LPM after the label-based lookup.

#### L7 userspace proxies

There is no CIDR awareness in the proxy implementations today, because they are primarily focused on L4+L7 rules, and Label-dependent L7 support is very new.  Currently, we push identity-based policies into proxies (NPDS), then push mappings from exact IPs to identities (NPHDS). If we want to support CIDR-dependent L4(+L7) policies in the proxy, we can either extend the NPDS to include CIDRs as an optional L3 dependency or alternatively extend NPHDS to include CIDR->identity mappings, then the actual policy match can remain the same. For performance and complexity reasons, the latter seems like a better option.

### Summary

The biggest concern for me right now is the inconsistency in policy. This PR implements approach 2.1 above in an attempt to make this more consistent, and I believe it fits better with the direction that Cilium is going with policy in general.

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

File: Documentation/policy/language.rst [hunk 1]
```diff
@@ -274,11 +274,11 @@ CIDR policies are used to define policies to and from endpoints which are not
 managed by Cilium and thus do not have labels associated with them. These are
 typically external services, VMs or metal machines running in particular
 subnets. CIDR policy can also be used to limit access to external services, for
-example to limit external access to a particular IP range.
+example to limit external access to a particular IP range. These policies never
+apply to traffic that comes from Cilium endpoints and is destined to Cilium
```
Comment (manalibhutiyani) [inline Documentation/policy/language.rst]: Do you want to also clarify,  if a specific CIDR range contains POD IPs, those will be ignored?

File: bpf/lib/lxc.h [hunk 2]
```diff
@@ -203,3 +202,4 @@ static inline bool __inline__ tc_index_skip_proxy(struct __sk_buff *skb)
 
 	return tc_index & TC_INDEX_F_SKIP_PROXY;
 }
+#endif
```
Comment (manalibhutiyani) [inline bpf/lib/lxc.h]: Nit: A /* corresponding #if */ would really help understand what is the corresponding #ifdef here.

File: bpf/lib/policy.h [hunk 3]
```diff
@@ -27,6 +27,10 @@
 #endif
 
 #ifdef REQUIRES_CAN_ACCESS
+static inline bool identity_is_reserved(__u32 identity)
+{
+	return identity < 256;
```
Comment (manalibhutiyani) [inline bpf/lib/policy.h]: Nit: Can we not hardcode this number and use a #define instead.

**Comment by joestringer** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: bpf/node_config.h [hunk 4]
```diff
@@ -58,7 +58,7 @@
 #define PROXY_MAP_SIZE 524288
 #define POLICY_MAP_SIZE 16384
 #define IPCACHE_MAP_SIZE 512000
-#define POLICY_PROG_MAP_SIZE 4096
+#define POLICY_PROG_MAP_SIZE ENDPOINTS_MAP_SIZE
```
Comment (tgraf) [inline bpf/node_config.h]: This is just a dummy header for compile testing so values don't really matter but it doesn't hurt

**Review by rlenglet — CHANGES_REQUESTED** [review]
I prefer to apply CIDR policies to all traffic. Because we'll have to also implement the same logic in Envoy filters, and it will be unnecessarily complex to do otherwise.

**Review by tgraf — APPROVED** [review]
Awesome. I 100% agree with this.

I think we should phrase it differently though:
CIDR rules apply if Cilium cannot map the source or destination to an identity derived based on endpoint labels, i.e. CIDR rules will apply to traffic classified as catch all identities such as from outside world, to outside world, from host, to host, from cluster, to cluster but not if the traffic can be classified with a source or destination of a [some name] identity.

We'll need to find a term for [some name] and make it clear in the docs what is a catch all identity such as host, world, cluster, proxy, ... and what are identities derived from endpoint labels/metadata.

**Comment by tgraf** [comment]
> I prefer to apply CIDR policies to all traffic. Because we'll have to also implement the same logic in Envoy filters, and it will be unnecessarily complex to do otherwise.

Can you be more specific? For Envoy it will be as simple as looking at the ipcache which is populated to Envoy already as I understand it to determine if there is a specific identity or not.

**Comment by jrajahalme** [comment]
@rlenglet Actually this makes NPDS/NPHDS implementation of CIDR-based L4/L7 easier as we can simply map CIDRs via (extended) NPHDS to a numeric ID and apply the policy configured via NPDS with no changes to NPDS required, as described by Joe in the last paragraph.

In the sidecar case there is a single mapping from remote IP to a numerical ID, like now, but extended to support CIDRs. I don't see the complication here.

File: bpf/node_config.h (hunk 4 above)
Comment (joestringer) [reply bpf/node_config.h]: `test/bpf/verifier-test.sh` fails without this change. It attempts to make reference beyond the size of the array, and the verifier fails it out. We could split the change out, this was for my local testing.

**Comment by rlenglet** [comment]
OK that sounds great.

**Review by rlenglet — APPROVED** [review]

**Comment by joestringer** [comment]
@tgraf thanks for the suggestion, I've rebased and updated the docs to this now:

> CIDR rules apply if Cilium cannot map the source or destination to an identity derived from endpoint labels, ie the Special Identities. For example, CIDR rules will apply to traffic where one side of the connection is:
>
> * A network endpoint outside the cluster
> * The host network namespace where the pod is running.
> * Within the cluster prefix but the IP’s networking is not provided by Cilium.

The "Special Identities" text links to the portion of the "Concepts" page which describes these identities.

**Comment by joestringer** [comment]
test-me-please

File: bpf/node_config.h (hunk 4 above)
Comment (joestringer) [reply bpf/node_config.h]: Shifted into #3316.

File: bpf/bpf_lxc.c [hunk 5]
```diff
@@ -209,15 +209,10 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 	/* If the packet is in the establishing direction and it's destined
 	 * within the cluster, it must match policy or be dropped. If it's
```
Comment (manalibhutiyani) [inline bpf/bpf_lxc.c:210]: I think this comment needs to be changed to reflect the new logic?

File: bpf/lib/policy.h [hunk 6]
```diff
@@ -133,11 +144,15 @@ policy_can_access_ingress(struct __sk_buff *skb, __u32 src_identity,
 	if (ret >= TC_ACT_OK)
 		return ret;
 
-	// cidr_addr_size is a compile time constant so this should all be inlined neatly.
-	if (cidr_addr_size == sizeof(union v6addr) && lpm6_ingress_lookup(cidr_addr))
-		goto allow;
-	if (cidr_addr_size == sizeof(__be32) && lpm4_ingress_lookup(*(__be32 *)cidr_addr))
-		goto allow;
+	/* CIDR policy only applies to traffic peering with something that is
+	 * not managed by Cilium. */
```
Comment (manalibhutiyani) [inline bpf/lib/policy.h:148]: Nit: `CIDR policy only applies to traffic peering with something that is not managed by Cilium. This translates to all the reserved identities.`

**Review by manalibhutiyani — APPROVED** [review]
Thanks for the awesome description. It really helped understand the code. I have added a few review comments which are mainly nits.

File: bpf/lib/policy.h (hunk 6 above)
Comment (tgraf) [reply bpf/lib/policy.h:148]: Good suggestions. I'm going to merge this so we can backport and tag the release today as that will take another couple of hours. Let's get these recommendations in in a follow-up PR.

**Pull request merged**